File: src/nbody_sim/python/simulation.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2024 nvidia-jetson-workload contributors

"""
Main simulation class for N-body simulation.
"""

from typing import Dict, List, Optional, Any, Tuple, Callable
import numpy as np
import time
import os

from src.nbody_sim.python.particle import ParticleSystem, Particle
from src.nbody_sim.python.integrator import Integrator


class NBodySimulation:
    """
    Main class for running an N-body simulation.
    
    This class combines a particle system with an integrator and provides
    methods to run the simulation and collect results.
    
    Attributes:
        system (ParticleSystem): The particle system being simulated
        integrator (Integrator): The integration method to use
        dt (float): Time step size
        duration (float): Total simulation time
        current_time (float): Current simulation time
        current_step (int): Current simulation step
        total_steps (int): Total number of steps to perform
        performance_data (Dict[str, Any]): Performance metrics
    """
    
    def __init__(
        self,
        system: ParticleSystem,
        integrator: Integrator,
        dt: float,
        duration: float
    ) -> None:
        """
        Initialize the simulation.
        
        Args:
            system: The particle system to simulate
            integrator: The integration method to use
            dt: Time step size
            duration: Total simulation time
        """
        self.system = system
        self.integrator = integrator
        self.dt = dt
        self.duration = duration
        
        # Initialize time tracking
        self.current_time = 0.0
        self.current_step = 0
        self.total_steps = int(duration / dt)
        
        # Initialize performance tracking
        self.performance_data = {
            "start_time": None,
            "end_time": None,
            "initial_energy": None,
            "final_energy": None,
        }
        
        # Initialize the integrator if needed
        self.integrator.initialize(system)
    
    def step(self) -> None:
        """
        Advance the simulation by one time step.
        """
        # Take a step with the chosen integrator
        self.integrator.step(self.system, self.dt)
        
        # Update time tracking
        self.current_time += self.dt
        self.current_step += 1
    
    def run(
        self, 
        callback: Optional[Callable[['NBodySimulation'], None]] = None,
        callback_interval: int = 1,
        record_initial_state: bool = True
    ) -> None:
        """
        Run the simulation for the specified duration.
        
        Args:
            callback: Optional function to call at each step
            callback_interval: How often to call the callback (in steps)
            record_initial_state: Whether to record initial energy for conservation checks
        """
        # Record start time
        self.performance_data["start_time"] = time.time()
        
        # Record initial energy if requested
        if record_initial_state:
            self.performance_data["initial_energy"] = self.system.total_energy()
        
        # Run the simulation
        for step in range(self.total_steps):
            # Take a step
            self.step()
            
            # Call the callback if provided and it's time to do so
            if callback is not None and step % callback_interval == 0:
                callback(self)
        
        # Record end time and energy
        self.performance_data["end_time"] = time.time()
        self.performance_data["final_energy"] = self.system.total_energy()
    
    def get_performance_metrics(self) -> Dict[str, float]:
        """
        Get performance metrics for the simulation.
        
        Returns:
            Dictionary of performance metrics
        """
        metrics = {}
        
        # Check if we have timing data
        if self.performance_data["start_time"] is not None and self.performance_data["end_time"] is not None:
            # Calculate elapsed time
            elapsed_time = self.performance_data["end_time"] - self.performance_data["start_time"]
            
            # Calculate steps per second
            if self.current_step > 0:
                steps_per_second = self.current_step / elapsed_time
            else:
                steps_per_second = 0.0
            
            metrics["total_time_ms"] = elapsed_time * 1000  # Convert to milliseconds
            metrics["steps_per_second"] = steps_per_second
        
        # Check if we have energy conservation data
        if self.performance_data["initial_energy"] is not None and self.performance_data["final_energy"] is not None:
            # Calculate relative energy error
            initial_energy = self.performance_data["initial_energy"]
            final_energy = self.performance_data["final_energy"]
            
            if initial_energy != 0:
                energy_error = abs((final_energy - initial_energy) / initial_energy)
            else:
                energy_error = abs(final_energy - initial_energy)
            
            metrics["initial_energy"] = initial_energy
            metrics["final_energy"] = final_energy
            metrics["energy_conservation_error"] = energy_error
        
        return metrics
    
    def save_state(self, filename: str) -> None:
        """
        Save the current state of the simulation to a file.
        
        Args:
            filename: Path to save the state file
        """
        # Extract particle data
        num_particles = self.system.num_particles
        positions = np.zeros((num_particles, 3))
        velocities = np.zeros((num_particles, 3))
        masses = np.zeros(num_particles)
        ids = np.zeros(num_particles, dtype=int)
        
        for i, particle in enumerate(self.system.particles):
            positions[i] = particle.position
            velocities[i] = particle.velocity
            masses[i] = particle.mass
            ids[i] = particle.particle_id
        
        # Save to file
        np.savez(
            filename,
            positions=positions,
            velocities=velocities,
            masses=masses,
            ids=ids,
            time=self.current_time,
            step=self.current_step,
            dt=self.dt,
            G=self.system.G
        )
    
    def load_state(self, filename: str) -> None:
        """
        Load a simulation state from a file.
        
        Args:
            filename: Path to the state file
        """
        # Load data from file
        data = np.load(filename)
        
        # Update simulation state
        self.current_time = float(data["time"])
        self.current_step = int(data["step"])
        
        # Update system and particles
        positions = data["positions"]
        velocities = data["velocities"]
        masses = data["masses"]
        ids = data["ids"]
        
        # Create new particles with loaded data
        particles = []
        for i in range(len(masses)):
            particles.append(
                Particle(
                    position=positions[i],
                    velocity=velocities[i],
                    mass=masses[i],
                    particle_id=int(ids[i])
                )
            )
        
        # Update the system
        self.system = ParticleSystem(particles, G=float(data["G"]))
        
        # Initialize the integrator with the new system
        self.integrator.initialize(self.system)
    
    def create_visualization_data(self, include_velocities: bool = True) -> Dict[str, Any]:
        """
        Create a data structure for visualization.
        
        Args:
            include_velocities: Whether to include velocity data
        
        Returns:
            Dictionary with visualization data
        """
        num_particles = self.system.num_particles
        
        # Extract data for visualization
        positions = np.zeros((num_particles, 3))
        masses = np.zeros(num_particles)
        ids = np.zeros(num_particles, dtype=int)
        
        for i, p in enumerate(self.system.particles):
            positions[i] = p.position
            masses[i] = p.mass
            ids[i] = p.particle_id
        
        # Create data structure
        vis_data = {
            "time": self.current_time,
            "step": self.current_step,
            "positions": positions.tolist(),
            "masses": masses.tolist(),
            "ids": ids.tolist(),
        }
        
        # Include velocities if requested
        if include_velocities:
            velocities = np.zeros((num_particles, 3))
            for i, p in enumerate(self.system.particles):
                velocities[i] = p.velocity
            vis_data["velocities"] = velocities.tolist()
        
        return vis_data

File: src/nbody_sim/cpp/src/simulation.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 Scott Friedman and Project Contributors

#include "nbody_sim/simulation.hpp"
#include <chrono>
#include <fstream>
#include <stdexcept>
#include <cmath>
#include <iomanip>
#include <sstream>

namespace nbody_sim {

Simulation::Simulation()
    : system_(std::make_unique<ParticleSystem>()),
      integrator_(std::make_unique<LeapfrogIntegrator>()),
      dt_(DEFAULT_TIMESTEP),
      duration_(10.0),
      current_time_(0.0),
      current_step_(0),
      total_steps_(static_cast<index_t>(duration_ / dt_)),
      initial_energy_(0.0),
      final_energy_(0.0) {}

Simulation::Simulation(
    std::unique_ptr<ParticleSystem> system,
    std::unique_ptr<Integrator> integrator,
    scalar_t dt,
    scalar_t duration
)
    : system_(std::move(system)),
      integrator_(std::move(integrator)),
      dt_(dt),
      duration_(duration),
      current_time_(0.0),
      current_step_(0),
      total_steps_(static_cast<index_t>(duration_ / dt_)),
      initial_energy_(0.0),
      final_energy_(0.0) {
    
    // Initialize the integrator
    integrator_->initialize(*system_);
}

void Simulation::step() {
    // Take a step with the integrator
    integrator_->step(*system_, dt_);
    
    // Update time tracking
    current_time_ += dt_;
    current_step_++;
}

void Simulation::run(StepCallback callback, index_t callback_interval) {
    // Record start time
    start_time_ = std::chrono::steady_clock::now();
    
    // Record initial energy
    initial_energy_ = system_->total_energy();
    
    // Run the simulation
    for (; current_step_ < total_steps_; ++current_step_) {
        // Take a step
        integrator_->step(*system_, dt_);
        
        // Update current time
        current_time_ += dt_;
        
        // Call the callback if provided and it's time to do so
        if (callback && current_step_ % callback_interval == 0) {
            callback(*this);
        }
    }
    
    // Record end time
    end_time_ = std::chrono::steady_clock::now();
    
    // Record final energy
    final_energy_ = system_->total_energy();
}

void Simulation::save_state(const std::string& filename) const {
    // Open file for writing
    std::ofstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Failed to open file for writing: " + filename);
    }
    
    // Write header with simulation parameters
    file.write(reinterpret_cast<const char*>(&current_time_), sizeof(current_time_));
    file.write(reinterpret_cast<const char*>(&current_step_), sizeof(current_step_));
    file.write(reinterpret_cast<const char*>(&dt_), sizeof(dt_));
    file.write(reinterpret_cast<const char*>(&duration_), sizeof(duration_));
    
    // Write integrator type
    IntegrationType integrator_type = integrator_->type();
    file.write(reinterpret_cast<const char*>(&integrator_type), sizeof(integrator_type));
    
    // Write gravitational constant
    scalar_t G = system_->gravitational_constant();
    file.write(reinterpret_cast<const char*>(&G), sizeof(G));
    
    // Write number of particles
    size_t num_particles = system_->size();
    file.write(reinterpret_cast<const char*>(&num_particles), sizeof(num_particles));
    
    // Write particle data
    for (size_t i = 0; i < num_particles; ++i) {
        const Particle& p = system_->particle(i);
        
        // Write position
        const Vec3& pos = p.position();
        file.write(reinterpret_cast<const char*>(&pos), sizeof(pos));
        
        // Write velocity
        const Vec3& vel = p.velocity();
        file.write(reinterpret_cast<const char*>(&vel), sizeof(vel));
        
        // Write mass
        scalar_t mass = p.mass();
        file.write(reinterpret_cast<const char*>(&mass), sizeof(mass));
        
        // Write ID
        index_t id = p.id();
        file.write(reinterpret_cast<const char*>(&id), sizeof(id));
    }
}

void Simulation::load_state(const std::string& filename) {
    // Open file for reading
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Failed to open file for reading: " + filename);
    }
    
    // Read header with simulation parameters
    file.read(reinterpret_cast<char*>(&current_time_), sizeof(current_time_));
    file.read(reinterpret_cast<char*>(&current_step_), sizeof(current_step_));
    file.read(reinterpret_cast<char*>(&dt_), sizeof(dt_));
    file.read(reinterpret_cast<char*>(&duration_), sizeof(duration_));
    
    // Update total_steps_ based on duration and dt
    total_steps_ = static_cast<index_t>(duration_ / dt_);
    
    // Read integrator type
    IntegrationType integrator_type;
    file.read(reinterpret_cast<char*>(&integrator_type), sizeof(integrator_type));
    
    // Create new integrator of the specified type
    integrator_ = Integrator::create(integrator_type);
    
    // Read gravitational constant
    scalar_t G;
    file.read(reinterpret_cast<char*>(&G), sizeof(G));
    
    // Read number of particles
    size_t num_particles;
    file.read(reinterpret_cast<char*>(&num_particles), sizeof(num_particles));
    
    // Create particles and add them to the system
    std::vector<Particle> particles;
    particles.reserve(num_particles);
    
    for (size_t i = 0; i < num_particles; ++i) {
        // Read position
        Vec3 pos;
        file.read(reinterpret_cast<char*>(&pos), sizeof(pos));
        
        // Read velocity
        Vec3 vel;
        file.read(reinterpret_cast<char*>(&vel), sizeof(vel));
        
        // Read mass
        scalar_t mass;
        file.read(reinterpret_cast<char*>(&mass), sizeof(mass));
        
        // Read ID
        index_t id;
        file.read(reinterpret_cast<char*>(&id), sizeof(id));
        
        // Create particle and add to vector
        particles.emplace_back(pos, vel, mass, id);
    }
    
    // Create new particle system
    system_ = std::make_unique<ParticleSystem>(particles, G);
    
    // Initialize the integrator with the new system
    integrator_->initialize(*system_);
}

std::map<std::string, double> Simulation::get_performance_metrics() const {
    std::map<std::string, double> metrics;
    
    // Calculate elapsed time
    if (end_time_ > start_time_) {
        const auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end_time_ - start_time_);
        metrics["total_time_ms"] = elapsed.count();
        
        // Calculate steps per second
        if (current_step_ > 0) {
            metrics["steps_per_second"] = 1000.0 * current_step_ / elapsed.count();
        }
    }
    
    // Calculate energy conservation
    if (initial_energy_ != 0.0) {
        metrics["initial_energy"] = initial_energy_;
        metrics["final_energy"] = final_energy_;
        metrics["energy_conservation_error"] = std::abs((final_energy_ - initial_energy_) / initial_energy_);
    }
    
    // Add system metrics
    metrics["particle_count"] = static_cast<double>(system_->size());
    metrics["current_time"] = current_time_;
    metrics["current_step"] = static_cast<double>(current_step_);
    
    return metrics;
}

std::map<std::string, std::vector<double>> Simulation::create_visualization_data(bool include_velocities) const {
    std::map<std::string, std::vector<double>> data;
    
    const size_t n = system_->size();
    
    // Initialize data arrays
    std::vector<double> positions_x(n);
    std::vector<double> positions_y(n);
    std::vector<double> positions_z(n);
    std::vector<double> masses(n);
    std::vector<double> ids(n);
    
    // Add position and mass data
    for (size_t i = 0; i < n; ++i) {
        const Particle& p = system_->particle(i);
        
        positions_x[i] = p.position().x;
        positions_y[i] = p.position().y;
        positions_z[i] = p.position().z;
        masses[i] = p.mass();
        ids[i] = static_cast<double>(p.id());
    }
    
    // Add data to result
    data["positions_x"] = positions_x;
    data["positions_y"] = positions_y;
    data["positions_z"] = positions_z;
    data["masses"] = masses;
    data["ids"] = ids;
    
    // Add velocity data if requested
    if (include_velocities) {
        std::vector<double> velocities_x(n);
        std::vector<double> velocities_y(n);
        std::vector<double> velocities_z(n);
        
        for (size_t i = 0; i < n; ++i) {
            const Particle& p = system_->particle(i);
            
            velocities_x[i] = p.velocity().x;
            velocities_y[i] = p.velocity().y;
            velocities_z[i] = p.velocity().z;
        }
        
        data["velocities_x"] = velocities_x;
        data["velocities_y"] = velocities_y;
        data["velocities_z"] = velocities_z;
    }
    
    // Add metadata
    std::vector<double> metadata = {
        static_cast<double>(n),
        current_time_,
        static_cast<double>(current_step_),
        dt_,
        duration_,
        system_->gravitational_constant()
    };
    
    data["metadata"] = metadata;
    
    return data;
}

std::unique_ptr<Simulation> Simulation::create_random_simulation(
    size_t num_particles,
    IntegrationType integration_type,
    scalar_t dt,
    scalar_t duration,
    unsigned int seed
) {
    // Create random particle system
    auto system = ParticleSystem::create_random_system(
        num_particles,
        10.0,  // box_size
        1.0,   // max_mass
        0.1,   // max_velocity
        1.0,   // G
        seed
    );
    
    // Create integrator
    auto integrator = Integrator::create(integration_type);
    
    // Create and return simulation
    return std::make_unique<Simulation>(
        std::move(system),
        std::move(integrator),
        dt,
        duration
    );
}

std::unique_ptr<Simulation> Simulation::create_solar_system_simulation(
    IntegrationType integration_type,
    scalar_t dt,
    scalar_t duration,
    scalar_t scale_factor
) {
    // Create solar system
    auto system = ParticleSystem::create_solar_system(scale_factor);
    
    // Create integrator
    auto integrator = Integrator::create(integration_type);
    
    // Create and return simulation
    return std::make_unique<Simulation>(
        std::move(system),
        std::move(integrator),
        dt,
        duration
    );
}

std::unique_ptr<Simulation> Simulation::create_galaxy_simulation(
    size_t num_particles,
    IntegrationType integration_type,
    scalar_t dt,
    scalar_t duration,
    unsigned int seed
) {
    // Create galaxy model
    auto system = ParticleSystem::create_galaxy_model(
        num_particles,
        10.0,  // radius
        1.0,   // height
        0.1,   // min_mass
        1.0,   // max_mass
        1.0,   // G
        seed
    );
    
    // Create integrator
    auto integrator = Integrator::create(integration_type);
    
    // Create and return simulation
    return std::make_unique<Simulation>(
        std::move(system),
        std::move(integrator),
        dt,
        duration
    );
}

} // namespace nbody_sim